### PR TITLE
chore: migrate to kcd

### LINKED ISSUE #499: Migrate CI to GitHub Actions
### Describe the feature you'd like:

The queue times on travis-ci are just getting stupidly long at this point and I gave GitHub Actions a go in a personal project recently and it was really nice. I'd like to convert this project over to using it instead.

### Suggested implementation:

Read the docs, implement the config.

### Describe alternatives you've considered:

There's plenty of other CIs out there, but I don't see any reason to consider other alternatives at this stage. Happy to hear suggestions though.

### Teachability, Documentation, Adoption, Migration Strategy:

N/A

### PR DESCRIPTION
**Pull request by marcosvega91**
**What**:

In this PR I have migrate the codebase to use `kcd-scripts` 🥳 
I have also added these main functionalities from kcd template generator (`yo kcd-oss`):

1. GitHub integration
2. Typescript compatibility
3. Semantic release

I have also used prettier/linter rules from kcd, but I have disabled some rules to not change the actual code.

close #499 #501 

**Why**:

In this way we already have typescript integration, GitHub pipeline and semantic release

**How**:

**Checklist**:

- [ ] Documentation updated
- [ ] Tests
- [ ] Ready to be merged
      
- [ ] Added myself to contributors table

### CODE REVIEW

**Review by mpeyper — CHANGES_REQUESTED** [review]
This is fantastic, thank you so much!

Things I'm sad about:
1. Moving tests into `src`
2. `lib` -> `dist`

Both of these are 100% legit changes that align much better with the community at large but are not my personal preference (I'm not suggesting changing them just for me, just calling out my own need to let some of this stuff go).

I might have missed it, but it doesn't look like you've added yourself as as a contributor in the README? You've absolutely deserved it with this change! I'll let you determine which categories you want to include because there's tonnes that fit for this.

Other than that there's a few comments to look at, but nothing too major. Overall this is a fantastic effort! Thanks again!

**Comment by mpeyper** [comment]
I see travis-ci is failing... I'm not worried about this as the point it to replace it, but we may want to delete the `.travis.yml` file in PR as well?

File: package.json [hunk 1]
```diff
@@ -23,80 +23,37 @@
...
     "prepare": "npm run build",
-    "build": "babel --out-dir lib src",
-    "format": "prettier-eslint --write \"**/*.{js,ts,json,yml,md,mdx}\"",
+    "build": "kcd-scripts build",
+    "lint": "kcd-scripts lint",
     "coverage": "codecov",
-    "test": "jest",
-    "test:ci": "npm run test && npm run coverage",
+    "test": "kcd-scripts test",
+    "test:update": "npm test -- --updateSnapshot --coverage",
```
[the remark this answers is not in the document]
Comment (marcosvega91) [reply package.json:33]: We can drop that script for now. We can add in feature if we need

File: package.json [hunk 2]
```diff
@@ -23,80 +23,37 @@
...
     "docs:dev": "docz dev",
     "docs:build": "docz build",
     "contributors:add": "all-contributors add"
   },
   "dependencies": {
-    "@babel/runtime": "^7.5.4",
-    "@types/testing-library__react-hooks": "^3.4.0"
```
[the remark this answers is not in the document]
Comment (marcosvega91) [reply package.json:38]: ah yes right

File: .eslintrc [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./node_modules/kcd-scripts/eslint.js",
+  "rules": {
+    "max-lines-per-function": "off",
+    "no-constant-condition": "off",
+    "no-await-in-loop": "off",
+    "react-hooks/rules-of-hooks": "off",
```
[the remark this answers is not in the document]
Comment (marcosvega91) [reply .eslintrc:7]: I have turned off to maintain compatibility with the actual codebase. I'm pretty sure that the await is in the code and not test.

https://github.com/testing-library/react-hooks-testing-library/blob/e61fa808bb731c48ecb090935cf2e3be1513532f/src/asyncUtils.js#L53-L75

[the remark this answers is not in the document]
Comment (mpeyper) [reply .eslintrc:7]: Sorry, it was the rules of `react-hooks/rules-of-hooks` rule that I was referring to, not `no-await-in-loop` (I didn't realise this was actually something to be avoided?)

File: package.json [hunk 4]
```diff
@@ -23,80 +23,36 @@
...
-    "@babel/core": "7.12.9",
-    "@babel/plugin-proposal-object-rest-spread": "7.12.1",
-    "@babel/plugin-transform-modules-commonjs": "7.12.1",
-    "@babel/plugin-transform-runtime": "7.12.1",
-    "@babel/preset-env": "7.12.7",
-    "@babel/preset-react": "7.12.7",
+    "@types/jest": "^26.0.15",
     "all-contributors-cli": "6.19.0",
     "babel-eslint": "10.1.0",
     "babel-plugin-module-resolver": "4.0.0",
```
Comment (mpeyper) [inline package.json:44]: This can be removed now too

File: .eslintrc (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcosvega91) [reply .eslintrc:7]: in this case I think that the `await` is necessary. In general we can push into an array the promise and wait everything with a `Promise.all`

`react-hooks/rules-of-hooks` is only for tests yes

File: package.json [hunk 5]
```diff
@@ -23,80 +23,36 @@
...
-    "@babel/cli": "7.12.8",
-    "@babel/core": "7.12.9",
-    "@babel/plugin-proposal-object-rest-spread": "7.12.1",
-    "@babel/plugin-transform-modules-commonjs": "7.12.1",
-    "@babel/plugin-transform-runtime": "7.12.1",
-    "@babel/preset-env": "7.12.7",
-    "@babel/preset-react": "7.12.7",
+    "@types/jest": "^26.0.15",
     "all-contributors-cli": "6.19.0",
     "babel-eslint": "10.1.0",
```
Comment (mpeyper) [inline package.json:43]: Does this need to be explicitly listed still?

File: package.json [hunk 6]
```diff
@@ -23,80 +23,37 @@
...
-    "prettier-eslint-cli": "5.0.0",
+    "kcd-scripts": "^7.2.0",
     "react": "17.0.1",
-    "react-test-renderer": "17.0.1"
+    "react-test-renderer": "17.0.1",
+    "typescript": "^4.1.2"
   },
   "peerDependencies": {
     "react": ">=16.9.0",
     "react-test-renderer": ">=16.9.0"
```
[the remark this answers is not in the document]
Comment (marcosvega91) [reply package.json:56]: I think that we can simply use * for both dependencies like we do with `@testing-library/react`. If we want we can also update the pipeline and build the library with different versions of react to check if also compatible with not released versions ( `next`, `experimental` )

**Comment by marcosvega91** [comment]
> I see travis-ci is failing... I'm not worried about this as the point it to replace it, but we may want to delete the `.travis.yml` file in PR as well?

Yes, I have removed the `travis` configuration. To make semantic release works you also need to update the settings of the project and add `NPM_TOKEN` and `GITHUB_TOKEN` into secrets

File: package.json (hunk 5 above)
Comment (marcosvega91) [reply package.json:43]: right we don't need yet

File: package.json [hunk 7]
```diff
@@ -26,7 +26,7 @@
     "setup": "npm install && npm run validate -s",
     "validate": "kcd-scripts validate",
     "prepare": "npm run build",
-    "build": "kcd-scripts build",
+    "build": "kcd-scripts build --out-dir lib",
```
Comment (marcosvega91) [inline package.json:29]: I have used `lib` folder instead of `dist` 😄

File: jest.config.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+const { jest: jestConfig } = require('kcd-scripts/config')
+
+module.exports = Object.assign(jestConfig, {
+  roots: ['<rootDir>/src', '<rootDir>/test'],
+  testMatch: ['<rootDir>/test/*.js']
+})
```
Comment (marcosvega91) [inline jest.config.js:6]: I have moved tests in test folder as it was before

File: .eslintrc (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpeyper) [reply .eslintrc:7]: Yeah ok. After this is merged I might look into turning the hook rules back on and fixing the issues/more granular disabling of it.

File: jest.config.js (hunk 8 above)
Comment (mpeyper) [reply jest.config.js:6]: This wasn't necessary, but thank you 😃

File: package.json (hunk 7 above)
Comment (mpeyper) [reply package.json:29]: Also not necessary, but again, thank you.

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (mpeyper) [reply package.json:55]: Unfortunately hooks are not available in any version of react, and specifically the async variant of `act` which we use directly, is only available from `16.9.x`, so I think `*` is a bit too permissive for us. Maybe the `>=` is fine then and it hasn't been an issue yet. I just sorry that a breaking change will come along one day and we won't be ready for it.

Don't worry to change anything for this PR.

[the remark this answers is not in the document]
Comment (marcosvega91) [reply package.json:55]: ah yes right, mine was a stupid observation hahahahah

**Comment by mpeyper** [comment]
> To make semantic release works you also need to update the settings of the project and add `NPM_TOKEN` and `GITHUB_TOKEN` into secrets

Thanks, I'll do that before merging (which I think we're ready to do now?)

**Review by mpeyper — APPROVED** [review]

**Comment by marcosvega91** [comment]
I think yes that we are ready. I hope that the pipeline will work 😄

**Comment by tigerabrodi** [comment]
@marcosvega91 Fun times :fire:, once this is through, we can start with the migration of render hooks to TypeScript right :heart_eyes:?

**Comment by marcosvega91** [comment]
Yes this is the base to make to library compatible with typescript 😄

**Comment by mpeyper** [comment]
> > To make semantic release works you also need to update the settings of the project and add `NPM_TOKEN` and `GITHUB_TOKEN` into secrets
> 
> Thanks, I'll do that before merging (which I think we're ready to do now?)

So, apparently `GITHUB_TOKEN` is automatically provided to actions and you aren't allowed to create secrets starting with `GITHUB_...`.  Also, `NPM_TOKEN` is set up at the org level so I didn't need to create that one either.

Ok, merging this now 🤞

**Pull request merged**